src/coredump/coredump.c

```diff
@@ -431,25 +407,88 @@ static int save_external_coredump(
+                 * amount of time, since if the storage is on tmpfs it will be charged against coredump's
+                 * memory accounting.
+                 * This is attempted in a best-effort fashion, in case anything goes wrong we log and
+                 * carry on. The uncompressed core is also useful only for journal storage and backtrace
+                 * generation, so only do that if either of these is enabled. */
+                if ((arg_storage == COREDUMP_STORAGE_JOURNAL && uncompressed_size <= arg_journal_size_max) ||
+                    uncompressed_size <= arg_process_size_max) {
+                        fd = open_tmpfile_linkable(fn, O_RDWR|O_CLOEXEC, &tmp);
+                        if (fd < 0)
+                                log_warning_errno(fd,
+                                                  "Failed to create temporary file for coredump %s, will not extract backtrace: %m",
+                                                  fn);
+                        else {
+                                r = decompress_stream(fn_compressed, fd_compressed, fd, max_size);
+                                if (r < 0) {
+                                        log_warning_errno(r,
+                                                          "Failed to decompress coredump %s, will not extract backtrace: %m",
+                                                          fn);
+                                        fd = safe_close(fd);
+                                }
+                                (void) unlink(tmp);
+                        }
+                }
+
+                *ret_filename = TAKE_PTR(fn_compressed);       /* compressed */
+                *ret_node_fd = TAKE_FD(fd_compressed);         /* compressed */
+                *ret_compressed_size = (uint64_t) st.st_size;  /* compressed */
+                *ret_data_fd = TAKE_FD(fd);                    /* uncompressed or closed */
+                *ret_size = uncompressed_size;                 /* uncompressed */
 
                 return 0;
 
         fail_compressed:
                 if (tmp_compressed)
                         (void) unlink(tmp_compressed);
+                return r;
         }
 
 uncompressed:
 #endif
 
+        /* If compression is disabled at build time or runtime, then just stream the core
+         * file from STDIN to the storage directory. */
+
+        fd = open_tmpfile_linkable(fn, O_RDWR|O_CLOEXEC, &tmp);
+        if (fd < 0) {
+                r = fd;
+                log_error_errno(fd, "Failed to create temporary file for coredump %s: %m", fn);
+                goto fail;
+        }
+
+        r = copy_bytes(input_fd, fd, max_size, 0);
+        if (r < 0) {
+                log_error_errno(r, "Cannot store coredump of %s (%s): %m",
+                                context->meta[META_ARGV_PID], context->meta[META_COMM]);
+                goto fail;
+        }
+        *ret_truncated = r == 1;
```

Comment:
hmm, it's a bit weird this is updated so early, given we can still fail. this is against our coding style really. please fix this while at it, i.e. store in a temporary variable `truncated` first, and then write to `ret_truncated` on success